packages/components/src/presentation-components/properties/inputs/UseNavigationPropertyTargetsLoader.ts

```diff
@@ -42,53 +50,82 @@ export interface NavigationPropertyTargetsResult {
 export interface UseNavigationPropertyTargetsLoaderProps {
   imodel: IModelConnection;
   ruleset?: Ruleset;
+  filterText?: string;
+  initialSelectedTarget?: string;
 }
 
 /** @internal */
 export function useNavigationPropertyTargetsLoader(props: UseNavigationPropertyTargetsLoaderProps) {
-  const { imodel, ruleset } = props;
-  const loadTargets = useCallback(
-    async (filter: string, loadedOptionsCount: number): Promise<NavigationPropertyTargetsResult> => {
-      if (!ruleset) {
-        return { options: [], hasMore: false };
+  const { imodel, ruleset, filterText, initialSelectedTarget } = props;
+
+  const [itemsLoader, setItemsLoader] = useState<NavigationPropertyItemsLoader | undefined>();
+
+  const [state, setLoadedOptions] = useState<NavigationPropertyValuesLoaderState>({
+    options: [],
+    isLoading: false,
+  });
+
+  // Get initial loader and values
+  useEffect(() => {
+    setLoadedOptions({ options: [], isLoading: false });
+    if (!ruleset) {
+      return;
+    }
+
+    const loader = new NavigationPropertyItemsLoader(
+      () => {
+        setLoadedOptions((prev) => ({ ...prev, isLoading: true }));
+      },
+      (newItems) => {
+        setLoadedOptions((prev) => ({
+          options: [...prev.options, ...newItems],
+          isLoading: false,
+        }));
+      },
+      async (filter?: string) => getItems(imodel, ruleset, filter),
+    );
+
+    void loader.loadItems(initialSelectedTarget);
+    setItemsLoader(loader);
+    return () => {
+      loader.dispose();
+    };
+  }, [imodel, initialSelectedTarget, ruleset]);
+
+  // On filter text change, check if more items need to be loaded
+  useEffect(() => {
+    if (!itemsLoader) {
+      return;
+    }
+
+    const timeout = setTimeout(() => {
+      if (itemsLoader?.needsLoadingItems(filterText)) {
```

Comment:
I think `?` is not needed anymore. Also `needsLoadingItems` could be moved into `loadItems` same as with unique values loader.

Reply:
Removed `needsLoadingItems`